### PR TITLE
Remove [most] CT variables from environment unless specify --permissive

### PR DESCRIPTION
**Pull request by rickporter-tuono**

### CODE REVIEW

**Review by jeking3 — APPROVED** [review]

**Review by nirvdrum — COMMENTED** [review]
I'm unclear on what the use case for the `--permissive` flag is. I think just disallowing overrides for the application's configuration is acceptable. Otherwise, you're bound to have confusing, difficult-to-debug behavior. And someone may use `--permissive` for `CT_SERVER_URL` only to be surprised when another team member changes `CT_ENV` to `"production"`.

File: src/config/mod.rs [hunk 1]
```diff
@@ -32,13 +32,18 @@ pub const ENV_VAR_PREFIX: &str = "CT_";
...
 
 // Environment variable name used to override the default server URL.
-#[allow(dead_code)]
 pub const CT_SERVER_URL: &str = "CT_SERVER_URL";
 
+// Environment variable name used to set the environment name.
+pub const CT_ENV: &str = "CT_ENV";
+
+// List of variables to remove to make a clean environment.
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (jeking3) [reply src/config/mod.rs:44]: While I may have originally pushed for the notion of configuration by environment (in addition to command line and config file), because that's what many CLIs do, in this case I think that may have been a mistake.

Doppler explicitly does not use environment variables to control their CLI, likely to avoid the issues we've been trying to solve here.  I believe it would be cleaner overall if we did not allow for any environment variable based configuration of our CLI, since we have a product that mucks with environment variables, and require only (command-line arguments | configuration file).  Inevitably we're going to end up with a customer who has a slew of CT_ prefixed variables that have nothing to do with us, AND they will use CT_ENV and CT_PROJECT (or worse, CT_API_KEY) which will mean nothing useful to us, and we'll eventually have wished we did not allow the CLI to be controlled by environment variables at all. :)

In other words, if you could not use CT_ anything to control our CLI, --permissive would not be necessary.  The none/overlay/underlay/exclusive control we put in earlier is still valid, however.

**Review by jeking3 — COMMENTED** [review]
Based on my reply I'm changing my approval to a comment.

If this goes in, it doesn't hurt anything.

In general, I'm starting to believe that allowing the CLI to be configured by environment variable will be a bug farm and we should restrict it to command-line and configuration file, and remove --permissive if we add it (because it will become unnecessary).

File: src/config/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dio-ct) [reply src/config/mod.rs:44]: In this case of a GitLabs CI/CD pipeline I think it using a sourced environment key stored as a variable.  That is how I have my test GitLab pipeline setup.  It is specifically using ``CT_API_KEY`` with the API token.  While the name collision is a concern I believe we do want to keep them for integration potentials.  We could pass in the key to the cmd line with -k in the pipeline but wouldn't you prefer the global?  I agree with Kevin on blocking the ``CT_`` prefix and expanding if required later.

[the remark this answers is not in the document]
Comment (nirvdrum) [reply src/config/mod.rs:44]: If the `CT_` prefix is problematic, we could use something less likely to conflict. `CT_CLI_` or `CLOUDTRUTH_` or `CLOUDTRUTH_CLI_`. The reason we used the short name was to operate similarly to the AWS suite of tools. Both the CLI and the client libraries will source the same set of variables. But, we don't have any client binding libraries, so using a more specific name would work. We'd need to add a warning or some such so existing users know to cut over.

The reason for configuration via env is mostly what @dio-ct indicated. It's also the dominant (or only) way of configuring tools run in Docker, Kubernetes, ECS, etc. We could force people to consume the values via CLI args (e.g., `--api-key ${SECRET_API_KEY}`), but I'm not sure that materially changes things.

**Review by jeking3 — APPROVED** [review]
This is impacted by #24 which is likely to go in, and after talking to Darryl and Rick about environment variables used by Doppler and the use cases they solve, I think this is a good tradeoff.  I think it makes sense for CLOUDTRUTH_PROJECT, CLOUDTRUTH_ENVIRONMENT (wow, that's pretty long now that I type it in... :) ) to always be passed through to the run shell. as those are probably useful.. and --permissive allows the token and host to go through.
